### DISCUSSION TITLE
Recommendation for splitting TRPC into multiple 'services' in monorepo

### DISCUSSION THREAD

**Discussion by Tobjoern** [trpc/trpc#4545] [Q&A]
I'm developing a semi-complex application, which features:
- a single database (postgres, prisma)
- one next.js app for 'Stakeholder A'
- one next.js app for 'Stakeholder B' (admin dashboard)
- one React Native app for 'Stakeholder C'
- Everything is stiched together using a turborepo monorepo

Until this point, I only had a single big 'api' TRPC project, which would be hosted through a single next.js app,
the other projects for different stakeholders, would simply call the api routes that were appropriate for them.

Approach A:
However, because of the growing level of complexity, I decided to split the big api project into 3 different ones, each of them with a separate TRPC router (created with 'createTRPCRouter').

My idea now would be, to host each of these packages in a separate next.js project, i.e. they are completely disconnect.

However, I'm wondering if this is the best way, to do so, as these packages still share the same database, and there may be additional burden / issues associated with having multiple projects accessing the same database, even if the db's types always stay in sync, because of the monorepo archietecture. (I'm not entirely sure about this point, feel free to disagree!)

Approach B:
The alternative approach would be to create one next.js project, which imports all of the different api-packages, and hosts them under separate endpoints, i.e.:
/api
   /stakeholder-a/trpc/[trpc].ts
   /stackeholder-b/trpc/[trpc].ts
   /...

Currently I'm leaning more towards A, as I'm not even sure, if approach B would be desireable, as multiple TRPC instances would be greated in a single next.js app (or is the limitation once per package?)
I'm sure I'm not the only person facing this issue, is there some other recommended approach I didn't consider yet?

**Comment by KATT** [comment]
I'd do a single API, but you can also merge the APIs client-side: https://github.com/trpc/trpc/tree/main/examples/soa

**Reply by KATT** [reply]
You can look at cal.com for reference https://github.com/calcom/cal.com

**Reply by Tobjoern** [reply]
Thanks a lot for the quick answer!
Cal.com's approach definitely looks interesting!

So to summarize, your suggestion would be to create a single 'api' next.js package, which merges the imported routes, creates a single trpc instance, and hosts it.
Would I then still be able to create separate trpc clients, with appropriate typings from the individual packages?

Because, I'd prefer the typings to be separated (i.e. so that I'm not able to call an 'admin' route, from the 'client' next.js project)

**Reply by Tobjoern** [reply]
All in all I'm trying to set some type boundaries in place,
having the routes 'trpc.stakeholderAAauth.login' and 'trpc.stakeholderBAuth.login', I'd prefer to have separate trpc client objects,
like 'trpc.auth.login' (in stakeholder A next.js package) and 'trpc.auth.login' (in stakeholder B next.js package) -> Where the respective API routes are called correctly.

**Comment by codersaadi** [comment]
those who are still searching can see 
https://github.com/codersaadi/turborepo-shadcn
https://github.com/codersaadi/advanced-starter-pack
